Add Packit configuration

nforro opened the pull request:
Hello @pbrezina, I'm opening this PR in response to https://github.com/packit/packit-service/issues/2372.

Packit configuration I've added includes both upstream and downstream jobs.

As for upstream, there is a `copr_build` job that builds authselect in Copr in the specified chroots on every pull request. You can see the results [here](https://github.com/nforro/authselect/pull/1/checks) on my fork. To enable it in this repository, please follow our [onboarding guide](https://packit.dev/docs/guide#how-to-set-up-packit-on-github) and install the *Packit-as-a-Service* GitHub application.

As for downstream, there is a `propose_downstream` job that updates authselect in dist-git when a new GitHub release is published and opens a pull request with the changes. I've only enabled this for rawhide, but it can be configured for stable branches as well. Once the opened dist-git PR is merged, a Koji build is automatically triggered and when finished automatic Bodhi update is created for rawhide.
There are however a few catches here. First, the spec file should be maintained upstream because the one in dist-git would be overwritten on every release. I noticed a few differences between the two, namely the `License` field, but that should be easy to fix. Second, Packit doesn't support downstream patches. Once `propose_downstream` runs, any references to them will be removed from the spec file. Would that be an issue for you?

pbrezina commented:
Thank you.

> Hello @pbrezina, I'm opening this PR in response to [packit/packit-service#2372](https://github.com/packit/packit-service/issues/2372).
> 
> Packit configuration I've added includes both upstream and downstream jobs.
> 
> As for upstream, there is a `copr_build` job that builds authselect in Copr in the specified chroots on every pull request. You can see the results [here](https://github.com/nforro/authselect/pull/1/checks) on my fork. To enable it in this repository, please follow our [onboarding guide](https://packit.dev/docs/guide#how-to-set-up-packit-on-github) and install the _Packit-as-a-Service_ GitHub application.

Ack.

> As for downstream, there is a `propose_downstream` job that updates authselect in dist-git when a new GitHub release is published and opens a pull request with the changes. I've only enabled this for rawhide, but it can be configured for stable branches as well. Once the opened dist-git PR is merged, a Koji build is automatically triggered and when finished automatic Bodhi update is created for rawhide.

Great, thank you.

> There are however a few catches here. First, the spec file should be maintained upstream because the one in dist-
git would be overwritten on every release. I noticed a few differences between the two, namely the `License` field, but that should be easy to fix. Second, Packit doesn't support downstream patches. Once `propose_downstream` runs, any references to them will be removed from the spec file. Would that be an issue for you?


I prefer to use the upstream spec file as the source of truth, although it it possible that not every recent change was added to upstream. I'll make sure to synchronize it before starting with packit.

However, how is changelog generated/managed? I don't want to synchronize changelog to the upstream project.

And last, currently, the spec file is populated with version and release from autoconf. But it looks that you just remove it. How does it work with packit?

pbrezina commented:
> Second, Packit doesn't support downstream patches. Once `propose_downstream` runs, any references to them will be removed from the spec file. Would that be an issue for you?

This is a problem since there are some changes for RHEL only that must be kept so it is synchronized correctly. Can we do something about it?

nforro commented:
> However, how is changelog generated/managed? I don't want to synchronize changelog to the upstream project.

By default Packit takes everything from the upstream spec file except for the `%changelog` section that is taken from the dist-git spec file and updated with a new entry that defaults to `- Update to version <version>`.

> This is a problem since there are some changes for RHEL only that must be kept so it is synchronized correctly. Can we do something about it?

The patches would have to be kept upstream and references to them present in the upstream spec file. Alternatively, I think they could be downloaded from dist-git in an action and references inserted into the upstream spec file before syncing it.

nforro commented:
> And last, currently, the spec file is populated with version and release from autoconf. But it looks that you just remove it. How does it work with packit?

By default, with `copr_build`, Packit updates `Version` with a version derived from the latest git tag and `Release` with a timestamp and git committish (to ensure that newer builds actually have higher NVRs). `propose_downstream` updates `Version` with a version derived from the git tag corresponding to the GitHub release that triggered it and resets `Release` to `1`.

nforro commented:
Hi, I've updated the configuration and added a `propose_downstream` job with an action that removes the RHEL10 downstream patches. You can see it here on my fork: https://github.com/nforro/authselect/blob/master/.packit.yaml
And here is the resulting dist-git PR: https://src.fedoraproject.org/rpms/authselect/pull-request/23

pbrezina commented:
> The generated project name is `packit/$namespace-$repository-$pr_id`, so you would run `dnf copr enable packit/authselect-authselect-$pr_id`. It can also be configured to build in an existing project (or even both).

Is it possible to build it inside authselect copr space?  https://copr.fedorainfracloud.org/groups/g/authselect/coprs/

I think this looks good. I'll read the documentation for the configuration file, but I guess it's all that is needed. If I get it correctly, everytime I do a new github release, it opens pull request for fedora-rawhide and once it is build, it creates koji build and submits it to bodhi?

nforro commented:
> Is it possible to build it inside authselect copr space? https://copr.fedorainfracloud.org/groups/g/authselect/coprs/

Yes, it should be possible by setting `owner: @authselect` on the `copr_build` job. Packit should automatically ask for permissions once the first build is triggered, but I have a feeling that only works for static projects, so you would probably have to allow Packit to create new projects in the group manually.

> I think this looks good. I'll read the documentation for the configuration file, but I guess it's all that is needed. If I get it correctly, everytime I do a new github release, it opens pull request for fedora-rawhide and once it is build, it creates koji build and submits it to bodhi?

You have to merge the pull request first, but yes. Koji builds and Bodhi updates can also be configured to trigger on non-Packit PR merges or direct pushes, but with the proposed configuration they will only trigger on a Packit PR merge.

pbrezina commented:
> > Is it possible to build it inside authselect copr space? https://copr.fedorainfracloud.org/groups/g/authselect/coprs/
> 
> Yes, it should be possible by setting `owner: @authselect` on the `copr_build` job. Packit should automatically ask for permissions once the first build is triggered, but I have a feeling that only works for static projects, so you would probably have to allow Packit to create new projects in the group manually.

Cool. Can we do this?

> > I think this looks good. I'll read the documentation for the configuration file, but I guess it's all that is needed. If I get it correctly, everytime I do a new github release, it opens pull request for fedora-rawhide and once it is build, it creates koji build and submits it to bodhi?
> 
> You have to merge the pull request first, but yes. Koji builds and Bodhi updates can also be configured to trigger on non-Packit PR merges or direct pushes, but with the proposed configuration they will only trigger on a Packit PR merge.

Packit PR only is sufficient.

nforro commented:
> Cool. Can we do this?

Unfortunately, it turns out it's not possible. Currently there are only two possibilities, either it can be a dynamically created project in `packit` namespace, e.g. https://copr.fedorainfracloud.org/coprs/packit/systemd-systemd-32932/, or it can be a static project with configured name in configured namespace/group.

nforro commented:
I've updated the config, to recap:

- Copr builds triggered on PRs, built in `packit/authselect-authselect-${PR_ID}` for now
- propose downstream triggered on a new release, preserving downstream RHEL patches and creating a PR in `rawhide`
- Koji build for `rawhide` triggered on merge of Packit PR
- Bodhi update for `rawhide` created automatically (by Bodhi, not by Packit)

pbrezina commented:
In the end, downstream patches are not needed anymore as c10s is no longer synced from Fedora. We can drop that and merge.

nforro commented:
> We can drop that and merge.

Just one thing, it is still necessary to install the [Packit-as-a-Service](https://github.com/marketplace/packit-as-a-service) GitHub application and complete the [approval step](https://packit.dev/docs/guide#2-approval) for the integration to actually work.

pbrezina commented:
Thank you, let's merge it and see how it works.
